File: core-render-ohos/src/main/cpp/libohos_render/expand/components/input/KRTextAreaView.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "libohos_render/manager/KRKeyboardManager.h"
#include "libohos_render/utils/KRConvertUtil.h"
#include "libohos_render/utils/KRViewUtil.h"
#include "libohos_render/expand/components/input/KRTextAreaView.h"

constexpr char kLineHeight[] = "lineHeight";

bool KRTextAreaView::SetProp(const std::string &prop_key, const KRAnyValue &prop_value,
                             const KRRenderCallback event_call_back) {
    if (kuikly::util::isEqual(prop_key, kLineHeight)) {
        kuikly::util::UpdateTextAreaNodeLineHeight(GetNode(), prop_value->toFloat());
        return true;
    }
    return KRTextFieldView::SetProp(prop_key, prop_value, event_call_back);
}


void KRTextAreaView::UpdateInputNodePlaceholder(const std::string &propValue) {
    ArkUI_AttributeItem item = {.string = propValue.c_str()};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_PLACEHOLDER, &item);
}
void KRTextAreaView::UpdateInputNodePlaceholderColor(const std::string &propValue) {
    ArkUI_NumberValue value = {.u32 = kuikly::util::ConvertToHexColor(propValue)};
    ArkUI_AttributeItem item = {&value, sizeof(ArkUI_NumberValue)};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_PLACEHOLDER_COLOR, &item);
}
void KRTextAreaView::UpdateInputNodeColor(const std::string &propValue) {
    ArkUI_NumberValue preparedColorValue[] = {{.u32 = kuikly::util::ConvertToHexColor(propValue)}};
    ArkUI_AttributeItem colorItem = {preparedColorValue, sizeof(preparedColorValue) / sizeof(ArkUI_NumberValue)};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_FONT_COLOR, &colorItem);
}
void KRTextAreaView::UpdateInputNodeCaretrColor(const std::string &propValue) {
    ArkUI_NumberValue value = {.u32 = kuikly::util::ConvertToHexColor(propValue)};
    ArkUI_AttributeItem item = {&value, sizeof(ArkUI_NumberValue)};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_CARET_COLOR, &item);
}

void KRTextAreaView::UpdateInputNodeKeyboardType(const std::string &propValue) {
    ArkUI_TextAreaType type = ARKUI_TEXTAREA_TYPE_NORMAL;
    if (propValue == "number") {
        type = ARKUI_TEXTAREA_TYPE_NUMBER;
    }else if (propValue == "email") {
        type = ARKUI_TEXTAREA_TYPE_EMAIL;
    }

    ArkUI_NumberValue value[] = {{.i32 = type}};
    ArkUI_AttributeItem item = {value, sizeof(value) / sizeof(ArkUI_NumberValue)};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_TYPE, &item);
}

void KRTextAreaView::UpdateInputNodeMaxLength(int maxLength) {
    ArkUI_NumberValue value[] = {{.i32 = maxLength}};
    ArkUI_AttributeItem item = {value, sizeof(value) / sizeof(ArkUI_NumberValue)};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_INPUT_MAX_LENGTH, &item);
}

uint32_t KRTextAreaView::GetInputNodeSelectionStartPosition() {
    auto item = kuikly::util::GetNodeApi()->getAttribute(GetNode(), NODE_TEXT_AREA_TEXT_SELECTION);
    return item ? item->value[0].i32 : 0;
}
void KRTextAreaView::UpdateInputNodeSelectionStartPosition(uint32_t index) {
    std::array<ArkUI_NumberValue, 2> value = {{{.i32 = static_cast<int32_t>(index)}, {.i32 = static_cast<int32_t>(index)}}};
    ArkUI_AttributeItem item = {value.data(), value.size()};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_TEXT_SELECTION, &item);
}
void KRTextAreaView::UpdateInputNodePlaceholderFont(uint32_t font_size, ArkUI_FontWeight font_weight) {
    ArkUI_NumberValue fontWeight = {.i32 = font_weight};
    ArkUI_NumberValue tempStyle = {.i32 = ARKUI_FONT_STYLE_NORMAL};
    std::array<ArkUI_NumberValue, 3> value = {{{.f32 = static_cast<float>(font_size)}, tempStyle, fontWeight}};
    ArkUI_AttributeItem item = {value.data(), value.size()};
    auto node = GetNode();
    kuikly::util::GetNodeApi()->setAttribute(node, NODE_TEXT_AREA_PLACEHOLDER_FONT, &item);

    {
        ArkUI_NumberValue valueSize[] = {{.f32 = static_cast<float>(font_size)}};
        ArkUI_AttributeItem itemSize = {valueSize, sizeof(valueSize) / sizeof(ArkUI_NumberValue)};
        kuikly::util::GetNodeApi()->setAttribute(node, NODE_FONT_SIZE, &itemSize);
    }
    {
        ArkUI_NumberValue fontWeight = {.i32 = font_weight};
        ArkUI_NumberValue valueWeight[] = {fontWeight};
        ArkUI_AttributeItem itemWeight = {valueWeight, 1};
        kuikly::util::GetNodeApi()->setAttribute(node, NODE_FONT_WEIGHT, &itemWeight);
    }
}
void KRTextAreaView::UpdateInputNodeContentText(const std::string &text) {
    ArkUI_AttributeItem item = {.string = text.c_str()};
    kuikly::util::GetNodeApi()->setAttribute(GetNode(), NODE_TEXT_AREA_TEXT, &item);
}

File: core-render-android/src/main/java/com/tencent/kuikly/core/render/android/expand/component/KRTextAreaView.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.core.render.android.expand.component

import android.content.Context
import android.view.Gravity
import android.view.MotionEvent
import android.view.inputmethod.EditorInfo
import com.tencent.kuikly.core.render.android.css.ktx.toDpF
import com.tencent.kuikly.core.render.android.export.KuiklyRenderCallback

/**
 * 多行输入框
 */
class KRTextAreaView(context: Context, softInputMode: Int?) : KRTextFieldView(context, softInputMode) {

    init {
        gravity = Gravity.LEFT or Gravity.TOP
    }

    override fun call(method: String, params: String?, callback: KuiklyRenderCallback?): Any? {
        return when (method) {
            METHOD_GET_INNER_CONTENT_HEIGHT -> getContentHeight(callback)
            else -> super.call(method, params, callback)
        }
    }

    override fun setProp(propKey: String, propValue: Any): Boolean {
        return when (propKey) {
            RETURN_KEY_TYPE -> setReturnKeyType(propValue)
            else -> super.setProp(propKey, propValue)
        }
    }

    private fun getContentHeight(callback: KuiklyRenderCallback?) {
        callback?.invoke(mapOf(
            "height" to kuiklyRenderContext.toDpF(layout.height.toFloat())
        ))
    }

    private fun setReturnKeyType(value: Any): Boolean {
        /**
         * 安卓 EditText 组件，默认情况下 IMEAction 和多行模式有冲突
         * 即默认情况下，如果设置了 isSingleLine=false，再设置 IMEAction 会无效
         *
         * 多行文本输入框需要设置 IMEAction（即 ReturnKeyType）的场景中，实际由于
         * 换行键会被替换成 "go" "next" "search" "send" "done" 等 IMEAction 键，
         * 因此实际上是单行模式（文本无法换行）只是长文本折行显示而已。
         *
         * 因此此处的处理方式为：
         * 1. 设置 isSingleLine=true 改为单行输入模式
         * 2. setHorizontallyScrolling(false) 禁止横向滚动来启用折行展示
         * 3. maxLines = Integer.MAX_VALUE 设置最大行数，否则文本区域会被截断
         * 4. 设置 IMEAction 键盘动作按钮展示
         */
        val method = transformationMethod
        isSingleLine = true
        transformationMethod = method // isSingleLine = true时，会将\n替换成空格, 不符合多行输入框的表现，这里强制把transformationMethod设置回去
        maxLines = Integer.MAX_VALUE
        setHorizontallyScrolling(false)

        imeOptions = when (value as String) {
            "none" -> {
                EditorInfo.IME_ACTION_NONE
            }
            "search" -> {
                EditorInfo.IME_ACTION_SEARCH
            }
            "send" -> {
                EditorInfo.IME_ACTION_SEND
            }
            "done" -> {
                EditorInfo.IME_ACTION_DONE
            }
            "next" -> {
                EditorInfo.IME_ACTION_NEXT
            }
            "go" -> {
                EditorInfo.IME_ACTION_GO
            }
            "previous" -> {
                EditorInfo.IME_ACTION_PREVIOUS
            }
            else -> {
                EditorInfo.IME_NULL
            }
        }
        return true
    }

    companion object {
        const val VIEW_NAME = "KRTextAreaView"

        private const val RETURN_KEY_TYPE = "returnKeyType"

        const val METHOD_GET_INNER_CONTENT_HEIGHT = "getInnerContentHeight"
    }

    override fun initSingleLine() {
        isSingleLine = false
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        // 解决Scroller内TextArea内容无法滚动
        if (event.actionMasked == MotionEvent.ACTION_DOWN &&
            computeVerticalScrollRange() > computeVerticalScrollExtent()) {
            parent?.requestDisallowInterceptTouchEvent(true)
        }
        return super.onTouchEvent(event)
    }
}
